### PR TITLE
stdenv: substituteStream: deprecate `--replace` in favor of `--replace-{fail,warn,quiet}`

### PR DESCRIPTION
**Pull request by pbsds**
## Description of changes

~~This change also adds the  --optional escape hatch.
There is no way this gets merged before ZHF is over.~~

This PR adds `--replace-fail`, `--replace-warn` and `--replace-quiet` to `substituteStream` in `stdenv`.
`--replace` has been deprecated, and will emit a warning if used.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [x] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by pbsds** [comment]
Here is a ripgrep for `substituteStream\(\): WARNING` in build logs I dumped from hydra: https://gist.github.com/pbsds/93bfc2d964e8f85e26af3e0f0c76fb17

Fixes #248565

**Comment by pbsds** [comment]
After trying to build some packages i saw how deep `--replace` is used, so I redid the approach to be a gradual transition where `--replace` is only deprecated.

File: pkgs/stdenv/generic/setup.sh [hunk 1]
```diff
@@ -834,9 +834,24 @@ substituteStream() {
...
+                is_quiet=1
+                ;&
             --replace)
+                if [ -z "$has_warned_replace_deprecation" ]; then
+                    echo "substituteStream(): WARNING: '--replace' is deprecated, use --replace-{fail,warn,quiet}. (in $description)" >&2
+                    has_warned_replace_deprecation=1
+                fi
+                ;&
+            --replace-warn)
+                is_required=
```
Comment (AndersonTorres) [inline pkgs/stdenv/generic/setup.sh:852]: Is this an "unset"?

**Review by AndersonTorres — DISMISSED** [review]
Besides, LGTM

File: pkgs/stdenv/generic/setup.sh (hunk 1 above)
Comment (pbsds) [reply pkgs/stdenv/generic/setup.sh:852]: No, its setting it to empty, to trigger the `[ -z x ]` test.
Would you prefer this?
```suggestion
                is_required=""
```

Comment (AndersonTorres) [reply pkgs/stdenv/generic/setup.sh:852]: Yes. Shell script is known to be a very sensitive language.

**Review by AndersonTorres — COMMENTED** [review]
LGTM, but OfBorg is complaining a bit

**Comment by pbsds** [comment]
Looks like timeouts and lack of disk space

**Comment by AndersonTorres** [comment]
@ofborg build

**Comment by pbsds** [comment]
If this does not make the feature freeze, then a backport would be rather smart, to not cause massive backporting conflict once the fixes comes rolling in

**Comment by AndersonTorres** [comment]
@ofborg build uftp

**Comment by Mindavi** [comment]
I think I'd like to see replace being stricter in the future, but it might be a bit of a hassle for external nixpkgs users. I do like the idea of this though. Why is quiet added? Presumably you want some feedback on whether your command does something, or what use case do you have for quiet? (I see some value, but there's also a chance it'll be used more than needed, I feel)

**Comment by pbsds** [comment]
@ofborg eval

**Review by AndersonTorres — APPROVED** [review]

**Comment by AndersonTorres** [comment]
Rebase.

**Review by Mindavi — APPROVED** [review]
Ok, I think this is a good idea conceptually

**Comment by Mindavi** [comment]
Let's go for this once ofborg eval succeeds (please ping me)

**Comment by RaitoBezarius** [comment]
> Let's go for this once ofborg eval succeeds (please ping me)

OfBorg eval passed but `stdenv` is still evaling on `aarch64-*` FYI

**Comment by katrinafyi** [comment]
Is there a bug where --replace-quiet will show the deprecation warning? The ;& in switch/case is a fall through.

**Comment by misuzu** [comment]
```
substituteStream(): ERROR: Invalid command line argument: --replace-fail
```
Any chance this would be backported to 23.11?

**Comment by AndersonTorres** [comment]
Do it.

**Pull request merged**